### PR TITLE
feat: add readonly flag

### PR DESCRIPTION
**Pull request by wcrum**
**Please check below, if the PR fulfills these requirements:**
- [x] Commit(s) and code follow the repositories guidelines.
- [x] Test(s) have been added or updated to support these change(s).
- [x] Doc(s) have been added or updated to support these change(s).

**Associated Links:**

- https://github.com/hauler-dev/hauler/issues/201

**Types of Changes:**

- Defaults running registry to be readonly, adds `--readonly` to default readonly behavior.

**Proposed Changes:**

- Additional `--readonly` flag.

**Verification/Testing of Changes:**

- Add content to hauler store.
- Start hauler registry with no additional flags, see behavior of registry is able to push to. Tear down.
- Start hauler registry with `--readonly` flag, see behavior which denies push to registry.

**Additional Context:**

- See https://github.com/hauler-dev/hauler/issues/201
> Secure by default... :D

### CODE REVIEW

**Comment by zackbradys** [comment]
Hey @wcrum, thank you for contributing and taking the time to put this together! It's been on our list for awhile.

I updated your pull request to change the language from `Writeable` to `ReadOnly`. We wanted to keep the same values as the configuration file and get this merged for `v1.0.5` this week.

File: cmd/hauler/cli/store/serve.go [hunk 1]
```diff
@@ -35,6 +34,7 @@ func (o *ServeRegistryOpts) AddFlags(cmd *cobra.Command) {
 	f.IntVarP(&o.Port, "port", "p", 5000, "Port to listen on.")
 	f.StringVar(&o.RootDir, "directory", "registry", "Directory to use for backend.  Defaults to $PWD/registry")
 	f.StringVarP(&o.ConfigFile, "config", "c", "", "Path to a config file, will override all other configs")
+	f.BoolVar(&o.ReadOnly, "readonly", true, "Run the registry as readonly.")
```
Comment (wcrum) [inline cmd/hauler/cli/store/serve.go:37]: Default value should be false due to !o.ReadOnly negating Bool.

Current behavior is where `--readonly=false` makes the registry **actually** readonly, while `--readonly` and no flag makes its writable.

**Review by wcrum — COMMENTED** [review]
Changing terminology from writable to readable cause opposite expected behavior.

**Comment by wcrum** [comment]
Do you want me to update https://github.com/hauler-dev/hauler-docs?

File: cmd/hauler/cli/store/serve.go [hunk 2]
```diff
@@ -136,9 +134,11 @@ func (o *ServeRegistryOpts) defaultRegistryConfig() *configuration.Configuration
 		Storage: configuration.Storage{
 			"cache":      configuration.Parameters{"blobdescriptor": "inmemory"},
 			"filesystem": configuration.Parameters{"rootdirectory": o.RootDir},
-
-			// TODO: Ensure this is toggleable via cli arg if necessary
-			// "maintenance": configuration.Parameters{"readonly.enabled": false},
+			"maintenance": configuration.Parameters{
+				"readonly": map[interface{}]interface{}{
+					"enabled": !o.ReadOnly,
+				},
```
Comment (dweomer) [inline cmd/hauler/cli/store/serve.go]: iirc this should be embedded another level via `configuration.Parameter`, or am i crazy?

Comment (dweomer) [reply cmd/hauler/cli/store/serve.go]: yeah, `configuration.Parameters` is an explicit `type` (tho declared not at the top of the package/file 🙄): https://github.com/distribution/distribution/blob/v3.0.0-beta.1/configuration/configuration.go#L377

Comment (wcrum) [reply cmd/hauler/cli/store/serve.go]: I tried that and it just didnt work. See below,

https://github.com/distribution/distribution/blob/main/registry/handlers/app.go#L139-L150

Comment (dweomer) [inline cmd/hauler/cli/store/serve.go]:
```suggestion
			"maintenance": configuration.Parameters{"readonly":
				configuration.Parameters{"enabled": !o.ReadOnly},
			},
```

**Review by dweomer — CHANGES_REQUESTED** [review]

File: cmd/hauler/cli/store/serve.go (hunk 1 above)
Comment (zackbradys) [reply cmd/hauler/cli/store/serve.go:37]: It was intentional! We want to ensure the `registry` always has its content populated from the `store`. If users want to be able to write to the registry outside of `hauler` then we wanted to make sure they make the choice to set `readonly` to `false`

Comment (wcrum) [reply cmd/hauler/cli/store/serve.go:37]: `--readonly=false` makes the registry readonly* with current logic.

**Review by dweomer — DISMISSED** [review]
oops, committed my suggestion. well, that was the only request change so approving!

File: cmd/hauler/cli/store/serve.go [hunk 3]
```diff
@@ -136,9 +134,10 @@ func (o *ServeRegistryOpts) defaultRegistryConfig() *configuration.Configuration
 		Storage: configuration.Storage{
 			"cache":      configuration.Parameters{"blobdescriptor": "inmemory"},
 			"filesystem": configuration.Parameters{"rootdirectory": o.RootDir},
-
-			// TODO: Ensure this is toggleable via cli arg if necessary
-			// "maintenance": configuration.Parameters{"readonly.enabled": false},
+			"maintenance": configuration.Parameters{"readonly":
+				configuration.Parameters{"enabled": !o.ReadOnly},
+			},
```
Comment (dweomer) [inline cmd/hauler/cli/store/serve.go]:
```suggestion
```

**Comment by wcrum** [comment]
Again, with current logic this breaks due to distribution logic.

Command running `hauler store serve registry`

```
panic: readonly config key must contain additional keys

goroutine 1 [running]:
github.com/distribution/distribution/v3/registry/handlers.NewApp({0x108f7a5f8?, 0x140004e7e60?}, 0x140006b6700)
	/Users/williamcrum/go/pkg/mod/github.com/distribution/distribution/v3@v3.0.0-20221208165359-362910506bc2/registry/handlers/app.go:143 +0x1dec
github.com/distribution/distribution/v3/registry.NewRegistry({0x108f7a5f8?, 0x140004e7e60?}, 0x140006b6700)
	/Users/williamcrum/go/pkg/mod/github.com/distribution/distribution/v3@v3.0.0-20221208165359-362910506bc2/registry/registry.go:156 +0x114
github.com/rancherfederal/hauler/internal/server.NewRegistry({0x108f7a5f8?, 0x140004e7e60?}, 0x140006b6700)
	/Users/williamcrum/Coding/Github/hauler/internal/server/registry.go:20 +0x28
github.com/rancherfederal/hauler/cmd/hauler/cli/store.ServeRegistryCmd({0x108f7a5f8, 0x140004e7740}, 0x140004e8040, 0x0?)
	/Users/williamcrum/Coding/Github/hauler/cmd/hauler/cli/store/serve.go:66 +0x1b0
github.com/rancherfederal/hauler/cmd/hauler/cli.addStoreServeRegistry.func1(0x14000014f00?, {0x1023ac5b7?, 0x4?, 0x1023ac5bb?})
	/Users/williamcrum/Coding/Github/hauler/cmd/hauler/cli/store.go:141 +0x6c
github.com/spf13/cobra.(*Command).execute(0x14000014f00, {0x1400045e2c0, 0x2, 0x2})
	/Users/williamcrum/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:983 +0x840
github.com/spf13/cobra.(*Command).ExecuteC(0x1400004e900)
	/Users/williamcrum/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/williamcrum/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/spf13/cobra.(*Command).ExecuteContext(0x108f7a5c0?, {0x108f7a5f8?, 0x140004e7740?})
	/Users/williamcrum/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1032 +0x48
main.main()
	/Users/williamcrum/Coding/Github/hauler/cmd/hauler/main.go:29 +0x130
```

File: cmd/hauler/cli/store/serve.go (hunk 2 above)
Comment (dweomer) [reply cmd/hauler/cli/store/serve.go]: thank you @wcrum! apologies for the churn on this one, i goofed up your change but have a fix that effectively reverts to your original changes (plus a correction for boolean negation)

File: cmd/hauler/cli/store/serve.go [hunk 4]
```diff
@@ -136,9 +134,9 @@ func (o *ServeRegistryOpts) defaultRegistryConfig() *configuration.Configuration
 		Storage: configuration.Storage{
 			"cache":      configuration.Parameters{"blobdescriptor": "inmemory"},
 			"filesystem": configuration.Parameters{"rootdirectory": o.RootDir},
-
-			// TODO: Ensure this is toggleable via cli arg if necessary
-			// "maintenance": configuration.Parameters{"readonly.enabled": false},
+			"maintenance": configuration.Parameters{"readonly":
```
Comment (dweomer) [inline cmd/hauler/cli/store/serve.go]:
```suggestion
			"maintenance": configuration.Parameters{
				"readonly": map[any]any{"enabled": o.ReadOnly},
			},
```

File: cmd/hauler/cli/store/serve.go [hunk 5]
```diff
@@ -136,9 +134,11 @@ func (o *ServeRegistryOpts) defaultRegistryConfig() *configuration.Configuration
 		Storage: configuration.Storage{
 			"cache":      configuration.Parameters{"blobdescriptor": "inmemory"},
 			"filesystem": configuration.Parameters{"rootdirectory": o.RootDir},
-
-			// TODO: Ensure this is toggleable via cli arg if necessary
-			// "maintenance": configuration.Parameters{"readonly.enabled": false},
+			"maintenance": configuration.Parameters{
+				"readonly": map[any]any{"enabled": o.ReadOnly},
+			},
+				configuration.Parameters{"enabled": !o.ReadOnly},
```
Comment (dweomer) [inline cmd/hauler/cli/store/serve.go]:
```suggestion
```

File: cmd/hauler/cli/store/serve.go [hunk 6]
```diff
@@ -136,9 +134,10 @@ func (o *ServeRegistryOpts) defaultRegistryConfig() *configuration.Configuration
 		Storage: configuration.Storage{
 			"cache":      configuration.Parameters{"blobdescriptor": "inmemory"},
 			"filesystem": configuration.Parameters{"rootdirectory": o.RootDir},
-
-			// TODO: Ensure this is toggleable via cli arg if necessary
-			// "maintenance": configuration.Parameters{"readonly.enabled": false},
+			"maintenance": configuration.Parameters{
+				"readonly": map[any]any{"enabled": o.ReadOnly},
+			},
+			},
```
Comment (dweomer) [inline cmd/hauler/cli/store/serve.go]: why is the web editor such a pos
```suggestion
```

**Review by dweomer — APPROVED** [review]

**Comment by zackbradys** [comment]
> Do you want me to update https://github.com/hauler-dev/hauler-docs?

If you have the time then that would be great! We have a few things missing from the docs right now and on the todo list after this release.

**Pull request merged**